dsgrid/config/file_schema.py

```diff
@@ -0,0 +1,221 @@
+            return None
+
+        type_upper = data_type.upper()
+        if type_upper not in SUPPORTED_TYPES:
+            supported_data_types = sorted(SUPPORTED_TYPES)
+            msg = f"{data_type=} is not one of {supported_data_types=}"
+            raise ValueError(msg)
+        return type_upper
+
+
+class FileSchema(DSGBaseModel):
+    """Defines the format of a data file (CSV, JSON, Parquet)."""
+
+    path: str | None = Field(description="Path to the file. Must be assigned during registration.")
+    columns: list[Column] = Field(
+        default=[], description="Custom schema for the columns in the file."
+    )
+    ignore_columns: list[str] = Field(
+        default=[],
+        description="List of column names to ignore (drop) when reading the file.",
+    )
+
+    @model_validator(mode="after")
+    def check_consistency(self) -> Self:
+        if len(self.columns) > 1:
+            check_uniqueness((x.name for x in self.columns), "column names")
+
+        # Check that ignore_columns don't overlap with columns
+        column_names = {x.name for x in self.columns}
+        ignore_set = set(self.ignore_columns)
+        overlap = column_names & ignore_set
+        if overlap:
+            msg = f"Columns cannot be in both 'columns' and 'ignore_columns': {overlap}"
+            raise ValueError(msg)
+
+        return self
+
+    def get_data_type_mapping(self) -> dict[str, str]:
+        """Return the mapping of column to data type."""
+        return {x.name: x.data_type for x in self.columns if x.data_type is not None}
+
+
+def read_data_file(schema: FileSchema) -> DataFrame:
+    """Read a data file from a schema.
+
+    Parameters
+    ----------
+    schema : FileSchema
+        Schema defining the file path and column types.
+
+    Returns
+    -------
+    DataFrame
+        A Spark DataFrame containing the file data.
+    """
+    if schema.path is None:
+        msg = "File path is not assigned"
+        raise DSGInvalidDataset(msg)
+
+    path = Path(schema.path)
+    if not path.exists():
+        msg = f"{path} does not exist"
+        raise FileNotFoundError(msg)
+
+    expected_columns = {x.name for x in schema.columns}
+
+    match path.suffix:
+        case ".parquet":
+            df = read_parquet(path)
+        case ".csv":
+            column_schema = _get_column_schema(schema, DUCKDB_COLUMN_TYPES)
+            df = read_csv_duckdb(path, schema=column_schema)
+        case ".json":
+            df = read_json(path)
+        case _:
+            msg = f"Unsupported file type: {path.suffix}"
+            raise DSGInvalidDataset(msg)
+
+    actual_columns = set(df.columns)
+    diff = expected_columns.difference(actual_columns)
+    if diff:
+        msg = f"Expected columns {diff} are not in {actual_columns=}"
+        raise DSGInvalidDataset(msg)
+
+    df = _drop_ignored_columns(df, schema.ignore_columns)
+    renames = _get_column_renames(schema)
+    df = _rename_columns(df, renames)
```

Comment:
Right now it appears that unexpected columns (actual columns - expected columns) would ride along for free. Instead we could detect, log, and drop those columns, perhaps modifying schema.ignore_columns accordingly.

Reply:
Unexpected columns will fail. We wouldn't ever want to modify ignore_columns, even if we automatically dropped unexpected columns. The registered dataset wouldn't have the columns and we wouldn't want to change the user's dataset.json5.